extension/src/lib.rs

```diff
@@ -1,324 +1,109 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 
-use codec::Encode;
+use codec::Decode as _;
 use frame_support::{
-	dispatch::{GetDispatchInfo, PostDispatchInfo},
-	pallet_prelude::*,
-	traits::OriginTrait,
+	dispatch::{GetDispatchInfo, PostDispatchInfo, RawOrigin},
+	ensure,
+	traits::{Contains, OriginTrait},
+	weights::Weight,
+};
+pub use functions::{
+	Decode, Decodes, DispatchCall, Equals, Function, FunctionId, Matches, Processor, ReadState,
 };
-use frame_system::RawOrigin;
 use pallet_contracts::chain_extension::{
-	BufInBufOutState, ChainExtension, Environment, Ext, InitState, RetVal,
+	ChainExtension, InitState, Result, RetVal, RetVal::Converging,
 };
-use sp_core::crypto::UncheckedFrom;
+pub use pallet_contracts::chain_extension::{Environment, Ext, State};
+use sp_core::Get;
 use sp_runtime::{traits::Dispatchable, DispatchError};
-use sp_std::vec::Vec;
+use std::marker::PhantomData;
 
+mod decoding;
+mod functions;
+mod matching;
 #[cfg(test)]
 mod tests;
-mod v0;
-
-/// Logging target for categorizing messages from the Pop API extension module.
-const LOG_TARGET: &str = "pop-api::extension";
-
-const DECODING_FAILED_ERROR: DispatchError = DispatchError::Other("DecodingFailed");
-// TODO: issue #93, we can also encode the `pop_primitives::Error::UnknownCall` which means we do use
-//  `Error` in the runtime and it should stay in primitives. Perhaps issue #91 will also influence
-//  here. Should be looked at together.
-const DECODING_FAILED_ERROR_ENCODED: [u8; 4] = [255u8, 0, 0, 0];
-const UNKNOWN_CALL_ERROR: DispatchError = DispatchError::Other("UnknownCall");
-// TODO: see above.
-const UNKNOWN_CALL_ERROR_ENCODED: [u8; 4] = [254u8, 0, 0, 0];
-
-type ContractSchedule<T> = <T as pallet_contracts::Config>::Schedule;
-
-/// Type of the state reader.
-pub trait ReadState {
-	/// Query of the state read operations.
-	type StateQuery: Decode;
 
-	/// Check if a state query is allowed.
-	fn contains(c: &Self::StateQuery) -> bool;
-
-	/// Reads state using the provided query, returning the result as a byte vector.
-	fn read(read: Self::StateQuery) -> Vec<u8>;
-
-	/// Decodes parameters into state query.
-	fn decode(params: &mut &[u8]) -> Result<Self::StateQuery, DispatchError> {
-		decode_checked(params)
-	}
-}
+type Schedule<T> = <T as pallet_contracts::Config>::Schedule;
 
-/// Type of the dispatch call filter.
-pub trait CallFilter {
-	/// Query of the dispatch calls operations.
-	type Call: Decode;
-
-	/// Check if runtime call is allowed.
-	fn contains(t: &Self::Call) -> bool;
-}
-
-/// Pop API chain extension.
+/// A configurable chain extension.
 #[derive(Default)]
-pub struct ApiExtension<I>(PhantomData<I>);
-
-impl<T, I> ChainExtension<T> for ApiExtension<I>
+pub struct Extension<C: Config>(PhantomData<C>);
```

Comment:
Alternative is Tin's implementation, which implements the Config trait on the runtime. The benefit is that you dont have an additional type, but I feel this is logically better, as we are configuring the chain extension plugged into pallet-contracts, plugged into the runtime. It keeps it encapsulated/logically distinct.

Reply:
I agree that having a chain extension that implements the Config trait makes more sense.